Fix Numba decorated function docstrings

thisac opened the pull request:
**Context:**
The generated docstrings for functions/methods decorated with Numba's `@jit` decorator didn't render properly since:

1. The decorator obscured the signature to Sphinx.

2. `numba` and `numba.jit` were mocked out in`conf.py` causing `magicmock` to provide its docstring instead of `numba` (which actually provides the correct, original docstring).

**Description of the Change:**

1. An `autodoc-signature-process` function was added to `conf.py` that checks whether the function is a Numba object, and in that case retrieves the original signature and forwards it to Sphinx.

2. The mocked out functions/modules `numba` and `numba.jit` are removed from `conf.py`, letting the correct docstrings be rendered.

3. The mocked out functions/modules `numpy` and `numpy.dtype` are removed from `conf.py`, letting the correct dtype be rendered inside the signatures.

Also fixed a few docstrings that weren't rendered correctly due to missing line-breaks.

**Benefits:**
The docstrings and signatures now render as they should.

**Possible Drawbacks:**
I'm not sure exactly why `numpy` and `numba` were mocked out in `conf.py` (nor the other funtions/method), so I might have broken something on the way. The documentation rendered nicely locally though.

**Related GitHub Issues:**
N/A

josh146 commented on docs/conf.py:
```diff
@@ -36,16 +38,16 @@ class TypeMock(type):
-    'numpy',
-    'numpy.dtype',
+    # 'numpy',
+    # 'numpy.dtype',
     'scipy',
     'scipy.special',
     'scipy.optimize',
     'scipy.stats',
     'cython',
     'thewalrus.libwalrus',
-    'numba',
-    'numba.jit'
+    # 'numba',
+    # 'numba.jit'
     ]
```
There were originally several reasons why these modules are mocked out:

* There is a bug in Sphinx (it might be fixed now) where Sphinx-autodoc would mistakenly include NumPy function documentation in projects that simply import it. This happened in Strawberry Fields, hence why it is probably mocked out here. I think it's okay to un-mock it now.

* Mocking allows Sphinx to import the project and inspect docstrings without requiring the dependencies be installed. This is quite important for `'thewalrus.libwalrus'`, as we don't want to have to build the C++ code on the RTDs environment.

josh146 replied:
I've noticed that `scipy` and `numpy` are now in the `doc/requirements.txt`, so are being installed on read the docs. So maybe SciPy doesn't have to be mocked any more either.

josh146 commented on docs/conf.py:
```diff
@@ -338,6 +340,23 @@ class TypeMock(type):
 from custom_directives import CustomGalleryItemDirective
 
+
+def process_numba_signature(app, what, name, obj, options, signature, return_annotation):
+    if isinstance(obj, numba.targets.registry.CPUDispatcher):
+        original = obj.py_func
+        orig_sig = inspect.signature(original)
+
+        if (orig_sig.return_annotation) is inspect._empty:
+            ret_ann = None
+        else:
+            ret_ann = orig_sig.return_annotation.__name__
+
+        return (str(orig_sig), ret_ann)
+
+    return (signature, return_annotation)
+
+
 def setup(app):
+    app.connect('autodoc-process-signature', process_numba_signature)
```
Awesome that this works!

josh146 commented:
https://github.com/numba/llvmlite/blob/c4f5a085c9f9c7a192579b02a11347bfc2ac768b/setup.py#L25

Looks like llvmlite disables installation on readthedocs environments :(

josh146 commented:
@thisac: I updated the `.readthedocs.yml` file to use Python 3.6, and this avoids `llvmlite/setup.py` from running, since the binary wheel is available.

But RTDs then runs into a different error:

```
  File "/home/docs/checkouts/readthedocs.org/user_builds/hafnian/checkouts/fix_numba_decorator/docs/conf.py", line 345, in process_numba_signature
    if isinstance(obj, numba.targets.registry.CPUDispatcher):
AttributeError: module 'numba' has no attribute 'targets'
```

See the full build log here: https://readthedocs.org/projects/hafnian/builds/11086294/

thisac commented:
> @thisac: I updated the `.readthedocs.yml` file to use Python 3.6, and this avoids `llvmlite/setup.py` from running, since the binary wheel is available.
> 
> But RTDs then runs into a different error:
> 
> ```
>   File "/home/docs/checkouts/readthedocs.org/user_builds/hafnian/checkouts/fix_numba_decorator/docs/conf.py", line 345, in process_numba_signature
>     if isinstance(obj, numba.targets.registry.CPUDispatcher):
> AttributeError: module 'numba' has no attribute 'targets'
> ```
> 
> See the full build log here: https://readthedocs.org/projects/hafnian/builds/11086294/
> 
> Edit: could something have changed in the latest Numba release? What version of Numba do you have installed locally?

Thanks @josh146. I've fixed the `AttributeError: module 'numba' has no attribute 'targets'` by updating the requirements for Numba to `numba>=0.49.1`. I believe I could fix the mocking issue, although, as you mentioned, it might not be necessary, so I've left it as it is at the moment.

nquesada commented:
hey @thisac : This is looking pretty good. There is one minor possibly related issue and is that the modules `quantum` and `samples` are no longer listed in The Walrus API:

![image](https://user-images.githubusercontent.com/991946/82607754-63793700-9b87-11ea-8fa9-42b13933322b.png)

josh146 commented on docs/conf.py:
```diff
@@ -36,16 +38,16 @@ class TypeMock(type):
-    'scipy.special',
-    'scipy.optimize',
-    'scipy.stats',
+    # 'numpy',
+    # 'numpy.dtype',
+    # 'scipy',
+    # 'scipy.special',
+    # 'scipy.optimize',
+    # 'scipy.stats',
```
Safe to delete all of these :)

thisac replied:
Done!

josh146 commented on docs/requirements.txt:
```diff
@@ -3,6 +3,7 @@ sphinx==2.2.2
+numba
 numpy>=1.9
 sympy>=1.5.1
 repoze.lru>=0.7
```
You will need to add scipy here for the docs to build, now that Scipy is not mocked out!

josh146 replied:
```suggestion
repoze.lru>=0.7
scipy>=1.2.1
```
